test/inductor/test_inductor_freezing.py

```diff
@@ -229,6 +232,140 @@ def foo(mod, inp):
+            def get_example_inputs():
+                return (torch.rand(2, 3, 5, 5).to(self.device),)
+
+        from torch._inductor.compile_fx import compile_fx, compile_fx_inner
+
+        nconv = 0
+
+        def my_inner_compile(gm, example_inputs, *args, **kwargs):
+            out = compile_fx_inner(gm, example_inputs, *args, **kwargs)
+
+            nonlocal nconv
+            convs = [n for n in gm.graph.nodes if n.target == aten.convolution.default]
+            nconv += len(convs)
+            for conv in convs:
+                weight_node = conv.args[1]
+                weight_const_tensor = getattr(gm, weight_node.target)
+                self.assertTrue(
+                    weight_const_tensor.is_contiguous(memory_format=torch.channels_last)
+                )
+                self.assertTrue(
+                    weight_node.meta["val"].is_contiguous(
+                        memory_format=torch.channels_last
+                    )
+                )
+
+            return out
+
+        mod = torch.compile(
+            Model().eval().to(self.device),
+            backend=functools.partial(compile_fx, inner_compile=my_inner_compile),
+        )
+        inp = mod.get_example_inputs()
+        with torch.no_grad():
+            mod(*inp)
+
+        # Only check the assertion for CUDA.
+        # For CPU, we may get torch.ops.mkldnn._convolution_pointwise.default
+        # in the joint graph rather than torch.ops.aten.convolution.default.
+        # Currently we only handle aten.convolution.default in layout
+        # optimization. That's why the count may be 0 here for CPU.
+        if self.device == "cuda":
+            self.assertTrue(nconv == 1)
+
+    def test_redundant_clone_for_layout_convert(self):
+        class Model(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.conv = nn.Conv2d(
+                    3, 128, kernel_size=3, padding=1, stride=1, bias=False
+                )
+
+            def forward(self, x):
+                y = x + 1
+                return self.conv(x), y
+
+            @staticmethod
+            def get_example_inputs():
+                return (torch.rand(2, 3, 5, 5).to(self.device),)
+
+        mod = Model().eval().to(self.device)
+        inp = mod.get_example_inputs()
+        with torch.no_grad():
+            expected_outputs = mod(*inp)
+
+        num_same_stride = 0
+        num_diff_stride = 0
+        orig_inductor_force_stride = lowering.inductor_force_stride
+
+        def debug_inductor_force_stride(input_tensor, stride):
+            nonlocal num_same_stride, num_diff_stride
+            input_tensor.realize()
+            if tuple(input_tensor.get_stride()) == tuple(stride):
+                num_same_stride += 1
+            else:
+                num_diff_stride += 1
+            return orig_inductor_force_stride(input_tensor, stride)
+
+        @contextlib.contextmanager
+        def mock_inductor_force_stride():
```

Comment:
The tests here are great, but seem like a lot of work. I wonder how much of this could have been done with `run_and_get_code`, I think a lot. Or similarly, if we should add other lightweight mechanisms to access various parts of the compilation stack via tests. Like, `run_and_get_inductor_graph`, which would dump out the inductor graph.

Reply:
`run_and_get_code` is great but may not help much here because it's not easy to decide if we have redundant clone by looking at the generated triton code.

`run_and_get_inductor_graph` can be a good alternative to do some assertion in fx Graph level. But if there is no objection, I still want to do the assertion in lowering (which happens laster compared to `run_and_get_inductor_graph`). 

I can make it a common API to override a lowering function. This way, code used in this test can be reused in other places .